README.md
# dbml-to-sqlalchemy

generate Model Class SqlAlchemy from database model dbml


## Installation


    pip install dbml-to-sqlalchemy
        
Or

    git clone https://github.com/fraoustin/dbml-to-sqlalchemy.git
    cd dbml-to-sqlalchemy
    python setup.py install

You can load test by

    flake8 --ignore E501,E226,E128,F401
    python -m unittest discover -s tests


## Usage

for sqlalchemy only 

    import os
    from re import sub
    import sqlalchemy as db
    from sqlalchemy.orm import Session
    from pydbml import PyDBML

    from dbml_to_sqlalchemy import createModel
    from dbml_to_sqlalchemy import mymodel

    database_file = "sqlite://"
    engine = db.create_engine(database_file, echo=False)
    conn = engine.connect()
    metadata = db.MetaData()


    try:
        from sqlalchemy.orm import DeclarativeBase

        class Base(DeclarativeBase):
            metadata = metadata
    except Exception:
        # for sqlalchemy 1.4
        from sqlalchemy.orm import declarative_base
        Base = declarative_base()

    source = """
    Table user {
        id integer [pk, increment, note:'key of user']
        name string [default: 'me', note:'only name']
        Note: 'Stores user data'
    }

    Table post {
        id integer [pk, increment]
        user_id integer [ref: > user.id]
    }
    """

    parsed = PyDBML(source)


    User = createModel(parsed.tables[0], Base)
    Post = createModel(parsed.tables[1], Base)

    print(User.__doc__)
    print(Post.__doc__)

    metadata.create_all(engine)

    with Session(engine) as session:
        user = User(id=1)
        session.add_all([user, ])
        session.commit()
        post = Post(id=1, user_id=1)
        session.add_all([post, ])
        session.commit()
        mypost = session.scalars(db.select(Post)).all()[0]
        print(mypost.user.name)



for flask-sqlalchemy

    import os
    import logging
    from flask import Flask
    from flask_sqlalchemy import SQLAlchemy
    from pydbml import PyDBML
    from dbml_to_sqlalchemy import createModel, mymodel

    app = Flask(__name__)

    # db SQLAlchemy
    database_file = "sqlite://"
    app.config["SQLALCHEMY_DATABASE_URI"] = database_file
    app.config["SQLALCHEMY_TRACK_MODIFICATIONS"] = False
    db = SQLAlchemy()

    source = """
    Table user {
    id integer [primary key]
    username varchar
    role varchar
    created_at timestamp
    }
    """

    parsed = PyDBML(source)
    createModel(parsed.tables[0], db.Model)


    def model_to_dict(obj):
        return {c.name: getattr(obj, c.name) for c in obj.__table__.columns}


    @app.route('/me')
    def mepath():
        return model_to_dict(mymodel.User.query.filter_by(id=1).first()), 200


    if __name__ == "__main__":
        db.init_app(app)
        with app.app_context():
            db.create_all()
            me = mymodel.User(id=1, username='me')
            db.session.add_all([me, ])
            db.session.commit()
        app.logger.setLevel(logging.DEBUG)
        app.run(host='0.0.0.0', port=5000, debug=True)

## TODO

- manage view from https://github.com/jklukas/sqlalchemy-views (https://stackoverflow.com/questions/9766940/how-to-create-an-sql-view-with-sqlalchemy)
- voir la gestion des defaults

CHANGES.md
# Feature

- generate documentation
- TODO

# V. 0.9.2

- generate specific class for Enumarate

# V. 0.9.1

- add sample in README

# V. 0.9.0

- init repos

tests/test01_sqlalchemy_basic.py
import unittest
import sqlalchemy as db
import sqlalchemy.sql.sqltypes
import sqlalchemy.exc
from sqlalchemy.orm import Session
from pydbml import PyDBML

from dbml_to_sqlalchemy import createModel


class BasicTest(unittest.TestCase):
    """
        Class for Basic Unitaire Test for dbml_to_sqlalchemy
    """
    def setUp(self):
        database_file = "sqlite://"
        self.engine = db.create_engine(database_file, echo=False)
        self.conn = self.engine.connect()
        self.metadata = db.MetaData()

        try:
            from sqlalchemy.orm import DeclarativeBase

            class Base(DeclarativeBase):
                metadata = self.metadata
        except Exception:
            # for sqlalchemy 1.4
            from sqlalchemy.orm import declarative_base
            Base = declarative_base()
        self.Base = Base

    def test_basic(self):
        with Session(self.engine) as session:
            source = """
            Table user {
            id integer [primary key]
            }
            """
            parsed = PyDBML(source)
            User = createModel(parsed.tables[0], self.Base)
            self.assertEqual("User", User.__name__)
            self.assertEqual("user", User.__tablename__)
            self.assertTrue("id" in dir(User))
            self.assertEqual(User.id.type.__class__, sqlalchemy.sql.sqltypes.Integer)
            self.assertTrue(User.id.primary_key)
            self.metadata.create_all(self.engine)
            user = User(id=1)
            session.add_all([user, ])
            session.commit()
            self.assertEqual(session.scalars(db.select(User)).all()[0].id, 1)

    def test_typ_columns(self):
        with Session(self.engine):
            source = """
            Table user {
            id integer [primary key]
            cola varchar
            colb varchar(20)
            colc decimal(10,2)
            cold blob
            cole unknown
            }
            """
            parsed = PyDBML(source)
            User = createModel(parsed.tables[0], self.Base)
            self.assertEqual(User.id.type.__class__, sqlalchemy.sql.sqltypes.Integer)
            self.assertTrue(User.id.primary_key)
            self.assertEqual(User.cola.type.__class__, sqlalchemy.sql.sqltypes.VARCHAR)
            self.assertEqual(User.colb.type.__class__, sqlalchemy.sql.sqltypes.VARCHAR)
            self.assertEqual(User.colb.type.length, 20)
            self.assertEqual(User.colc.type.__class__, sqlalchemy.sql.sqltypes.DECIMAL)
            self.assertEqual(User.colc.type.precision, 10)
            self.assertEqual(User.colc.type.scale, 2)
            self.assertEqual(User.cold.type.__class__, sqlalchemy.sql.sqltypes.BLOB)
            self.assertEqual(User.cole.type.__class__, sqlalchemy.sql.sqltypes.String)

    def test_null_columns(self):
        with Session(self.engine) as session:
            source = """
            Table user {
            id integer [primary key]
            cola varchar [not null]
            colb varchar
            }
            """
            parsed = PyDBML(source)
            User = createModel(parsed.tables[0], self.Base)
            self.assertEqual(User.cola.nullable, False)
            self.assertEqual(User.colb.nullable, True)
            self.metadata.create_all(self.engine)
            user = User(id=1, cola='test', colb='test')
            session.add_all([user, ])
            self.assertEqual(session.commit(), None)
            user = User(id=2, cola='test', colb=None)
            session.add_all([user, ])
            self.assertEqual(session.commit(), None)
            with self.assertRaises(sqlalchemy.exc.IntegrityError):
                user = User(id=3, cola=None, colb='test')
                session.add_all([user, ])
                session.commit()

    def test_unique_columns(self):
        with Session(self.engine) as session:
            source = """
            Table user {
            id integer [primary key]
            cola varchar [unique]
            colb varchar
            }
            """
            parsed = PyDBML(source)
            User = createModel(parsed.tables[0], self.Base)
            self.assertEqual(User.cola.unique, True)
            self.assertEqual(User.colb.unique, False)
            self.metadata.create_all(self.engine)
            user = User(id=1, cola='test', colb='test')
            session.add_all([user, ])
            self.assertEqual(session.commit(), None)
            user = User(id=2, cola='test2', colb='test')
            session.add_all([user, ])
            self.assertEqual(session.commit(), None)
            with self.assertRaises(sqlalchemy.exc.IntegrityError):
                user = User(id=3, cola='test2', colb=None)
                session.add_all([user, ])
                session.commit()

    def test_default_columns(self):
        with Session(self.engine) as session:
            source = """
            Table user {
            id integer [primary key]
            cola varchar [default: "coucou"]
            colb boolean [default: True]
            colc integer [default: 2]
            }
            """
            parsed = PyDBML(source)
            User = createModel(parsed.tables[0], self.Base)
            self.metadata.create_all(self.engine)
            user = User(id=1, cola=None)
            session.add_all([user, ])
            session.commit()
            self.assertEqual(session.scalars(db.select(User)).all()[0].cola, "coucou")
            self.assertEqual(session.scalars(db.select(User)).all()[0].colb, True)
            self.assertEqual(session.scalars(db.select(User)).all()[0].colc, 2)

    def test_enum_columns(self):
        with Session(self.engine) as session:
            source = """
            Table user {
            id integer [primary key]
            cola job_status
            }

            enum job_status {
                created [note: 'Waiting to be processed']
                running
                done
                failure
            }
            """
            parsed = PyDBML(source)
            User = createModel(parsed.tables[0], self.Base)
            self.metadata.create_all(self.engine)
            user = User(id=1, cola=None)
            session.add_all([user, ])
            session.commit()
            self.assertEqual(session.scalars(db.select(User)).all()[0].cola, None)
            user = User(id=2, cola='running')
            session.add_all([user, ])
            session.commit()
            self.assertEqual(session.scalars(db.select(User)).all()[1].cola.name, "running")
            with self.assertRaises(sqlalchemy.exc.IntegrityError):
                user = User(id=3, cola="wrong_value")
                session.add_all([user, ])
                session.commit()

    def test_autoincrement_columns(self):
        with Session(self.engine) as session:
            source = """
            Table user {
            id integer [primary key, increment]
            }
            """
            parsed = PyDBML(source)
            User = createModel(parsed.tables[0], self.Base)
            self.metadata.create_all(self.engine)
            user = User()
            session.add_all([user, ])
            session.commit()
            self.assertEqual(session.scalars(db.select(User)).all()[0].id, 1)
            user = User()
            session.add_all([user, ])
            session.commit()
            self.assertEqual(session.scalars(db.select(User)).all()[1].id, 2)

    def test_pk_multiple(self):
        with Session(self.engine) as session:
            source = """
            Table user {
            id integer
            idd integer

            indexes {
                (id, idd) [pk]
            }
            }
            """
            parsed = PyDBML(source)
            User = createModel(parsed.tables[0], self.Base)
            self.metadata.create_all(self.engine)
            user = User(id=1, idd=1)
            session.add_all([user, ])
            session.commit()
            self.assertEqual(session.scalars(db.select(User)).all()[-1].id, 1)
            user = User(id=1, idd=2)
            session.add_all([user, ])
            session.commit()
            self.assertEqual(session.scalars(db.select(User)).all()[-1].id, 1)
            user = User(id=2, idd=2)
            session.add_all([user, ])
            session.commit()
            self.assertEqual(session.scalars(db.select(User)).all()[-1].id, 2)
            with self.assertRaises(sqlalchemy.exc.IntegrityError):
                user = User(id=2, idd=2)
                session.add_all([user, ])
                session.commit()

    def test_unique_multiple(self):
        with Session(self.engine) as session:
            source = """
            Table user {
            id integer [pk, increment]
            idd integer
            iddd integer

            indexes {
                (idd, iddd) [unique]
            }
            }
            """
            parsed = PyDBML(source)
            User = createModel(parsed.tables[0], self.Base)
            self.metadata.create_all(self.engine)
            user = User(idd=1, iddd=1)
            session.add_all([user, ])
            session.commit()
            self.assertEqual(session.scalars(db.select(User)).all()[-1].idd, 1)
            user = User(idd=1, iddd=2)
            session.add_all([user, ])
            session.commit()
            self.assertEqual(session.scalars(db.select(User)).all()[-1].idd, 1)
            user = User(idd=2, iddd=2)
            session.add_all([user, ])
            session.commit()
            self.assertEqual(session.scalars(db.select(User)).all()[-1].idd, 2)
            with self.assertRaises(sqlalchemy.exc.IntegrityError):
                user = User(idd=2, iddd=2)
                session.add_all([user, ])
                session.commit()


tests/test02_sqlalchemy_reference.py
import unittest
import sqlalchemy as db
import sqlalchemy.sql.sqltypes
import sqlalchemy.exc
from sqlalchemy.orm import Session
from pydbml import PyDBML

from dbml_to_sqlalchemy import createModel

from sqlalchemy.engine import Engine
from sqlalchemy import event


@event.listens_for(Engine, "connect")
def set_sqlite_pragma(dbapi_connection, connection_record):
    cursor = dbapi_connection.cursor()
    cursor.execute("PRAGMA foreign_keys=ON")
    cursor.close()


class BasicTest(unittest.TestCase):
    """
        Class for Basic Unitaire Test for dbml_to_sqlalchemy
    """
    def setUp(self):
        database_file = "sqlite://"
        self.engine = db.create_engine(database_file, echo=False)
        self.conn = self.engine.connect()
        self.metadata = db.MetaData()

        try:
            from sqlalchemy.orm import DeclarativeBase

            class Base(DeclarativeBase):
                metadata = self.metadata
        except Exception:
            # for sqlalchemy 1.4
            from sqlalchemy.orm import declarative_base
            Base = declarative_base()
        self.Base = Base

    def test_basic(self):
        with Session(self.engine) as session:
            source = """
            Table usera {
                id integer [pk]
            }

            Table posta {
                id integer [pk]
                userid integer [ref: > usera.id]
            }
            """
            parsed = PyDBML(source)
            User = createModel(parsed.tables[0], self.Base)
            Post = createModel(parsed.tables[1], self.Base)
            self.metadata.create_all(self.engine)
            user = User(id=1)
            session.add_all([user, ])
            session.commit()
            post = Post(id=1, userid=1)
            session.add_all([post, ])
            session.commit()
            self.assertEqual(session.scalars(db.select(Post)).all()[0].id, 1)
            self.assertEqual(session.scalars(db.select(Post)).all()[0].userid, 1)
            with self.assertRaises(sqlalchemy.exc.IntegrityError):
                post = Post(id=2, userid=2)
                session.add_all([post, ])
                session.commit()

    def test_basic_inv(self):
        with Session(self.engine) as session:
            source = """
            Table userb {
                id integer [pk, ref: < postb.userid]
            }

            Table postb {
                id integer [pk]
                userid integer
            }
            """
            parsed = PyDBML(source)
            User = createModel(parsed.tables[0], self.Base)
            Post = createModel(parsed.tables[1], self.Base)
            self.metadata.create_all(self.engine)
            user = User(id=1)
            session.add_all([user, ])
            session.commit()
            post = Post(id=1, userid=1)
            session.add_all([post, ])
            session.commit()
            self.assertEqual(session.scalars(db.select(Post)).all()[0].id, 1)
            self.assertEqual(session.scalars(db.select(Post)).all()[0].userid, 1)
            with self.assertRaises(sqlalchemy.exc.IntegrityError):
                post = Post(id=2, userid=2)
                session.add_all([post, ])
                session.commit()

    def test_multicolumn(self):
        with Session(self.engine) as session:
            source = """
            Table userc {
                id integer
                idd integer
            indexes {
                (id, idd) [pk]
            }
            }

            Table postc {
                id integer [pk]
                userid integer
                useridd integer
            }
            Ref: postc.(userid, useridd) > userc.(id, idd)
            """
            parsed = PyDBML(source)
            User = createModel(parsed.tables[0], self.Base)
            Post = createModel(parsed.tables[1], self.Base)
            self.metadata.create_all(self.engine)
            user = User(id=1, idd=1)
            session.add_all([user, ])
            session.commit()
            post = Post(id=1, userid=1, useridd=1)
            session.add_all([post, ])
            session.commit()
            self.assertEqual(session.scalars(db.select(Post)).all()[0].id, 1)
            self.assertEqual(session.scalars(db.select(Post)).all()[0].userid, 1)
            self.assertEqual(session.scalars(db.select(Post)).all()[0].useridd, 1)
            with self.assertRaises(sqlalchemy.exc.IntegrityError):
                post = Post(id=2, userid=2, useridd=1)
                session.add_all([post, ])
                session.commit()

    def test_relation_basic(self):
        with Session(self.engine) as session:
            source = """
            Table userd {
                id integer [pk, ref: < postd.userid]
            }

            Table postd {
                id integer [pk]
                userid integer
            }
            """
            parsed = PyDBML(source)
            User = createModel(parsed.tables[0], self.Base)
            Post = createModel(parsed.tables[1], self.Base)
            self.metadata.create_all(self.engine)
            user = User(id=1)
            session.add_all([user, ])
            session.commit()
            post = Post(id=1, userid=1)
            session.add_all([post, ])
            session.commit()
            self.assertEqual(session.scalars(db.select(Post)).all()[0].id, 1)
            self.assertEqual(session.scalars(db.select(Post)).all()[0].userid, 1)
            self.assertEqual(session.scalars(db.select(Post)).all()[0].userd.id, 1)
            post = Post(id=2, userid=1)
            session.add_all([post, ])
            session.commit()
            self.assertEqual(len(session.scalars(db.select(User)).all()[0].postds), 2)


tests/test01_flasksqlalchemy_basic.py
import unittest
from flask import Flask, request, json
from pydbml import PyDBML

from flask_sqlalchemy import SQLAlchemy
from dbml_to_sqlalchemy import createModel


class BasicTest(unittest.TestCase):
    """
        Class for Basic Unitaire Test for flask_sqlalchemy_api
    """
    def setUp(self):
        self.app = Flask(__name__)
        self.app.testing = True
        self.app.config["SQLALCHEMY_DATABASE_URI"] = "sqlite://"
        self.db = SQLAlchemy()

        self.db.init_app(self.app)

    def test_create_insert(self):
        with self.app.test_client():
            source = """
            Table user {
            id integer [primary key]
            }
            """
            parsed = PyDBML(source)
            User = createModel(parsed.tables[0], self.db.Model)
            with self.app.app_context():
                self.db.drop_all()
                self.db.create_all()
                item = User(id=1)
                self.db.session.add(item)
                self.db.session.commit()
                self.assertEqual(User.query.all()[0].id, 1)


sample/sample01.py
import os
from re import sub
import sqlalchemy as db
from sqlalchemy.orm import Session
from pydbml import PyDBML

from dbml_to_sqlalchemy import createModel
from dbml_to_sqlalchemy import mymodel

database_file = "sqlite://"
engine = db.create_engine(database_file, echo=False)
conn = engine.connect()
metadata = db.MetaData()


try:
    from sqlalchemy.orm import DeclarativeBase

    class Base(DeclarativeBase):
        metadata = metadata
except Exception:
    # for sqlalchemy 1.4
    from sqlalchemy.orm import declarative_base
    Base = declarative_base()

source = """
Table user {
    id integer [pk, increment, note:'key of user']
    name string [default: 'me', note:'only name']
    Note: 'Stores user data'
}

Table post {
    id integer [pk, increment]
    user_id integer [ref: > user.id]
}
"""

parsed = PyDBML(source)


User = createModel(parsed.tables[0], Base)
Post = createModel(parsed.tables[1], Base)

print(User.__doc__)
print(Post.__doc__)

metadata.create_all(engine)

with Session(engine) as session:
    user = User(id=1)
    session.add_all([user, ])
    session.commit()
    post = Post(id=1, user_id=1)
    session.add_all([post, ])
    session.commit()
    mypost = session.scalars(db.select(Post)).all()[0]
    print(mypost.user.name)


dbml_to_sqlalchemy/main.py
from re import sub, search
import enum
from dbml_to_sqlalchemy import mymodel

from pydbml import PyDBML
from sqlalchemy import PrimaryKeyConstraint, UniqueConstraint, ForeignKeyConstraint
from sqlalchemy.sql.schema import Column
from sqlalchemy.sql import expression
from sqlalchemy.orm import relationship
import sqlalchemy.types
from sqlalchemy import Enum
import sqlalchemy.sql.sqltypes
types = {typ.lower(): getattr(sqlalchemy.types, typ) for typ in dir(sqlalchemy.types) if typ in dir(sqlalchemy.sql.sqltypes) and '_' not in typ}

__version__ = '0.9.1'


def toCamelCase(st):
    st = sub('[^a-zA-Z0-9 \n]', '', st.replace('.', ' '))
    return sub(r"(_|-)", " ", st).title().replace(" ", "")


def toColumnCase(st):
    st = sub('[^a-zA-Z0-9 _\n]', '', st.replace('.', ' '))
    return st.lower().replace(" ", "_")


def toTableCase(st):
    return st.lower().replace(" ", "_")


def getType(st, default=sqlalchemy.types.String):
    if st.__class__.__name__ == 'Enum':
        return Enum
    return types.get(st.split('(')[0].lower(), default)


def getTypeParams(st, module):
    try:
        if st.__class__.__name__ == 'Enum':            
            setattr(module, st.name, enum.Enum(st.name, [item.name for item in st.items]))
            return [getattr(module, st.name), ], {"create_constraint": True}
        return [param.isnumeric() and int(param) or param for param in search(r'\((.*)\)', st).group(0)[1:-1].split(',')], {}
    except Exception:
        return [], {}


def getServerDefault(val):
    if isinstance(val, bool):
        if bool is False:
            return expression.false()
        return expression.true()
    if val is None:
        return val
    return str(val)


def spec_many(newt, name):
    @property
    def decorator(f):
        return [getattr(elt, name) for elt in getattr(f, newt)]
    return decorator


def spec_doc(col):
    doc = []
    if col.pk:
        doc.append('primary key')
    if col.autoinc:
        doc.append('autoincrement')
    if col.not_null:
        doc.append('not null')
    if col.unique:
        doc.append('unique')
    if col.default is not None:
        doc.append('default: %s' % col.default)
    if len(doc) > 0:
        return '(%s)' % ', '.join(doc)
    return ''


def createModel(table, *cls, module=mymodel):
    if len(table.note.text) == 0:
        table.note.text = '%s Table' % toCamelCase(table.name)
    cols = {col.name: Column(toColumnCase(col.name), getType(col.type)(*getTypeParams(col.type, module)[0], **getTypeParams(col.type, module)[1]), primary_key=col.pk, autoincrement=col.autoinc, nullable=not (col.not_null), default=col.default, server_default=getServerDefault(col.default), unique=col.unique, comment=col.note) for col in table.columns}
    tableArgs = []
    for index in [index for index in table.indexes if index.pk is True]:
        tableArgs.append(PrimaryKeyConstraint(*[cols[col.name] for col in index.subjects]))
    for index in [index for index in table.indexes if index.unique is True]:
        subargs = {}
        if index.name is not None:
            subargs = {'name': index.name}
        tableArgs.append(UniqueConstraint(*[cols[col.name] for col in index.subjects], **subargs))
    SomeClass = type(toCamelCase(table.name), cls, {
        "__tablename__": toTableCase(table.name),
        "__table_args__": tuple(tableArgs),
        "__doc__": "%s\n\n%s" % (table.note, '\n'.join([":param %s: %s %s\n:type %s: %s" % (col.name, col.note, spec_doc(col), col.name, col.type) for col in table.columns])),
        **cols
    })
    setattr(module, toCamelCase(table.name), SomeClass)
    for ref in [ref for ref in table.database.refs if (ref.table1 == table or ref.table2 == table) and getattr(module, toCamelCase(ref.col2[0].table.name), None) is not None and getattr(module, toCamelCase(ref.col1[0].table.name), None) is not None]:
        class1 = getattr(module, toCamelCase(ref.table1.name))
        class2 = getattr(module, toCamelCase(ref.table2.name))
        if ref.type in ('<'):
            class2.__table_args__ = class2.__table_args__ + (ForeignKeyConstraint([getattr(class2, col.name) for col in ref.col2], [getattr(class1, col.name) for col in ref.col1], name=ref.name), )
        if ref.type in ('>', '-'):
            class1.__table_args__ = class1.__table_args__ + (ForeignKeyConstraint([getattr(class1, col.name) for col in ref.col1], [getattr(class2, col.name) for col in ref.col2], name=ref.name), )
        if ref.type == '<':
            setattr(class1, "%ss" % class2.__name__.lower(), relationship(class2.__name__, back_populates=class1.__name__.lower()))
            class1.__doc__ = class1.__doc__ + "\n:param %ss:\n:type %ss: relationship(%s)" % (class2.__name__.lower(), class2.__name__.lower(), class2.__name__)
            setattr(class2, class1.__name__.lower(), relationship(class1.__name__, back_populates="%ss" % class2.__name__.lower()))
            class2.__doc__ = class2.__doc__ + "\n:param %s :\n:type %s: %s" % (class1.__name__.lower(), class1.__name__.lower(), class1.__name__)
        elif ref.type == '>':
            setattr(class1, class2.__name__.lower(), relationship(class2.__name__, back_populates="%ss" % class1.__name__.lower()))
            class1.__doc__ = class1.__doc__ + "\n:param %s:\n:type %s: %s" % (class2.__name__.lower(), class2.__name__.lower(), class2.__name__)
            setattr(class2, "%ss" % class1.__name__.lower(), relationship(class1.__name__, back_populates=class2.__name__.lower()))
            class2.__doc__ = class2.__doc__ + "\n:param %ss:\n:type %ss: relationship(%s):" % (class1.__name__.lower(), class1.__name__.lower(), class1.__name__)
        elif ref.type == '-':
            setattr(class1, class2.__name__.lower(), relationship(class2.__name__, uselist=False, back_populates=class1.__name__.lower()))
            setattr(class2, class1.__name__.lower(), relationship(class1.__name__, uselist=False, back_populates=class2.__name__.lower()))
        if ref.type == '<>':
            newt_name = "%s_%s" % (ref.table1.name, ref.table2.name)
            newt_col1 = "\n".join(["%s_%s %s" % (col.table.name, col.name, col.type) for col in ref.col1])
            newt_col2 = "\n".join(["%s_%s %s" % (col.table.name, col.name, col.type) for col in ref.col2])
            newt_pk = "%s, %s" % ("\n".join(["%s_%s" % (col.table.name, col.name) for col in ref.col1]), "\n".join(["%s_%s" % (col.table.name, col.name) for col in ref.col2]))
            new_table = "Table %s\n{\n%s\n%s\nindexes {\n(%s) [pk]\n}\n}" % (newt_name, newt_col1, newt_col2, newt_pk)
            newt = createModel(PyDBML(new_table).tables[0], *cls, module=module)
            newt.__table_args__ = newt.__table_args__ + (ForeignKeyConstraint([getattr(newt, "%s_%s" % (col.table.name, col.name)) for col in ref.col1], [getattr(class1, col.name) for col in ref.col1]), )
            newt.__table_args__ = newt.__table_args__ + (ForeignKeyConstraint([getattr(newt, "%s_%s" % (col.table.name, col.name)) for col in ref.col2], [getattr(class2, col.name) for col in ref.col2]), )
            setattr(class1, "%ss" % newt.__name__.lower(), relationship(newt.__name__, back_populates=class1.__name__.lower()))
            class1.__doc__ = class1.__doc__ + "\n:param %ss:\n:type %ss: relationship(%s)" % (newt.__name__.lower(), newt.__name__.lower(), newt.__name__)
            setattr(newt, class1.__name__.lower(), relationship(class1.__name__, back_populates="%ss" % newt.__name__.lower()))
            newt.__doc__ = newt.__doc__ + "\n:param %s:\n:type %s: %s" % (class1.__name__.lower(), class1.__name__.lower(), class1.__name__)
            setattr(class2, "%ss" % newt.__name__.lower(), relationship(newt.__name__, back_populates=class2.__name__.lower()))
            class2.__doc__ = class2.__doc__ + "\n:param %ss:\n:type %ss: relationship(%s):" % (newt.__name__.lower(), newt.__name__.lower(), newt.__name__)
            setattr(newt, class2.__name__.lower(), relationship(class2.__name__, back_populates="%ss" % newt.__name__.lower()))
            newt.__doc__ = newt.__doc__ + "\n:param %s:\n:type %s: %s:" % (class2.__name__.lower(), class2.__name__.lower(), class2.__name__)
            setattr(class1, "%ss" % class2.__name__.lower(), spec_many("%ss" % newt.__name__.lower(), class2.__name__.lower()))
            class1.__doc__ = class1.__doc__ + "\n:param %ss:\n:type %ss: relationship(%s)" % (class2.__name__.lower(), class2.__name__.lower(), class2.__name__)
            setattr(class2, "%ss" % class1.__name__.lower(), spec_many("%ss" % newt.__name__.lower(), class1.__name__.lower()))
            class2.__doc__ = class2.__doc__ + "\n:param %ss:\n:type %ss: relationship(%s)" % (class1.__name__.lower(), class1.__name__.lower(), class1.__name__)
            setattr(module, newt.__name__, newt)
    return getattr(module, toCamelCase(table.name))


sample/sample02.py
import os
import logging
from flask import Flask
from flask_sqlalchemy import SQLAlchemy
from pydbml import PyDBML
from dbml_to_sqlalchemy import createModel, mymodel

app = Flask(__name__)

# db SQLAlchemy
database_file = "sqlite://"
app.config["SQLALCHEMY_DATABASE_URI"] = database_file
app.config["SQLALCHEMY_TRACK_MODIFICATIONS"] = False
db = SQLAlchemy()

source = """
Table user {
id integer [primary key]
username varchar
role varchar
created_at timestamp
}
"""

parsed = PyDBML(source)
createModel(parsed.tables[0], db.Model)


def model_to_dict(obj):
    return {c.name: getattr(obj, c.name) for c in obj.__table__.columns}


@app.route('/me')
def mepath():
    return model_to_dict(mymodel.User.query.filter_by(id=1).first()), 200


if __name__ == "__main__":
    db.init_app(app)
    with app.app_context():
        db.create_all()
        me = mymodel.User(id=1, username='me')
        db.session.add_all([me, ])
        db.session.commit()
    app.logger.setLevel(logging.DEBUG)
    app.run(host='0.0.0.0', port=5000, debug=True)
